usr.bin/ldd/ldd.c

```diff
@@ -344,19 +348,85 @@ usage(void)
 	exit(1);
 }
 
+static bool
+check_notes(const char *buf, size_t len)
+{
+	const Elf_Note *note;
+	const char *name;
+	size_t namesz, descsz;
+
+	for (;;) {
+		if (len < sizeof(*note))
+			return (false);
+		note = (const Elf_Note *)(uintptr_t)buf;
+		buf += sizeof(*note);
+		len -= sizeof(*note);
+
+		namesz = roundup2(note->n_namesz, 4);
+		descsz = roundup2(note->n_descsz, 4);
+		if (len < namesz + descsz)
+			return (false);
+
+		name = buf;
+		if (strncmp(name, "FreeBSD", namesz) == 0 && 
+		    note->n_type == NT_FREEBSD_ABI_TAG)
+			return (true);
+
+		buf += namesz + descsz;
+		len -= namesz + descsz;
+	}
+
+	return (false);
+}
+
+static bool
+is_freebsd_elf(const char *fname, Elf *elf, GElf_Ehdr *ehdr)
+{
+	GElf_Shdr shdr;
+	Elf_Scn *scn;
+	Elf_Data *d;
+
+	switch (ehdr->e_ident[EI_OSABI]) {
+	case ELFOSABI_FREEBSD:
+		return (true);
+	case ELFOSABI_NONE:
+		break;
+	default:
+		return (false);
+	}
+
+	scn = NULL;
+	while ((scn = elf_nextscn(elf, scn)) != NULL) {
```

Comment:
We should be looking for PT_NOTEs, not SHT_NOTEs. Keep those section headers out of loaders :)

Reply:
Hmm, so libelf only does all its Elf_Data handling (where it byte swaps the Elf_Note structures for you) for sections, not for program segments.  To use PT_NOTE I'd have to do all the parsing of the notes by hand.

Reply:
Well, we expect the endianness to be the same for ldd and the target binary, no?

Reply:
Certainly for something we will dlopen().  I'd already switched on EI_DATA, but I could perhaps refuse to check notes in mismatched binaries at least.